pkg/virt-controller/services/template.go

```diff
@@ -22,27 +22,52 @@ package services
 import (
 	"path/filepath"
 	"strconv"
+	"strings"
 
-	kubev1 "k8s.io/api/core/v1"
+	k8sv1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
 
 	"kubevirt.io/kubevirt/pkg/api/v1"
 	"kubevirt.io/kubevirt/pkg/precond"
-	"kubevirt.io/kubevirt/pkg/registry-disk"
+	registrydisk "kubevirt.io/kubevirt/pkg/registry-disk"
 )
 
+const configMapName = "kube-system/virt-controller"
+const allowEmulationKey = "debug.allowEmulation"
+
 type TemplateService interface {
-	RenderLaunchManifest(*v1.VirtualMachine) *kubev1.Pod
+	RenderLaunchManifest(*v1.VirtualMachine) (*k8sv1.Pod, error)
 }
 
 type templateService struct {
 	launcherImage   string
 	virtShareDir    string
 	imagePullSecret string
+	store           cache.Store
+}
+
+func IsEmulationAllowed(store cache.Store) (bool, error) {
+	obj, exists, err := store.GetByKey(configMapName)
```

Comment:
I actually think that we could have used the volume api here (i.e. expose ConfigMap as a voluem to VM pod, let launcher take value from there **OR** expose ConfigMap as Env vars using DownardAPI, let the launcher query env var)

@rmohr @davidvossel thoughts?

Reply:
We agreed that virt-controller should just configure the pod the right way. I prefer to have the virt-launcher pod to be created with no surprises. E.g. using env-vars just hides that we allow emulation. Adding an extra flag is streight foward to understand. DownwardAPI and Env vars are very useful in general if you don't want to think about k8s on the application level (you simply want to configure and start your application, which has nothing to do with k8s. you don't want kubectl, client-go or so integrated, just to configure your app). So a good way for normal workloads. We are a k8s application we can do it this way too and it is even less surprising from that perspective.

Reply:
+1

Just wanted to throw that other idea into the round :)

Reply:
It's also possible the config-map could contain information in the future that we don't want visible to the virt-launcher pod.